File: 2402-meeting-rooms-iii/2402-meeting-rooms-iii.py
import heapq

class Solution:
    def mostBooked(self, n: int, meetings: list[list[int]]) -> int:
        # Sort meetings by start time
        meetings.sort()
        
        # Heap of available room numbers (min-heap)
        available = list(range(n))
        heapq.heapify(available)
        
        # Heap of ongoing meetings as (end_time, room_id)
        busy = []
        
        # Count of meetings handled per room
        count = [0] * n

        for start, end in meetings:
            duration = end - start
            
            # Free up rooms that are done before the current meeting starts
            while busy and busy[0][0] <= start:
                end_time, room = heapq.heappop(busy)
                heapq.heappush(available, room)

            if available:
                # Assign to the available room with the smallest ID
                room = heapq.heappop(available)
                heapq.heappush(busy, (end, room))
                count[room] += 1
            else:
                # No rooms available: delay meeting to the earliest possible time
                end_time, room = heapq.heappop(busy)
                new_end = end_time + duration
                heapq.heappush(busy, (new_end, room))
                count[room] += 1

        # Return the room with max count, preferring the smallest index
        max_meetings = max(count)
        return count.index(max_meetings)

File: 2402-meeting-rooms-iii/2402-meeting-rooms-iii.cpp
class Solution {
public:
    int mostBooked(int n, vector<vector<int>>& meetings) {
        sort(meetings.begin(), meetings.end());

        vector<int> count(n, 0); // Meeting count per room
        priority_queue<int, vector<int>, greater<>> available;
        for (int i = 0; i < n; ++i) available.push(i);

        // Min-heap: {endTime, roomId}
        priority_queue<pair<long long, int>, vector<pair<long long, int>>, greater<>> busy;

        for (auto& m : meetings) {
            long long start = m[0], end = m[1], duration = end - start;

            // Free up rooms that are done by current meeting's start
            while (!busy.empty() && busy.top().first <= start) {
                available.push(busy.top().second);
                busy.pop();
            }

            if (!available.empty()) {
                int room = available.top(); available.pop();
                busy.emplace(end, room);
                ++count[room];
            } else {
                auto [next_free_time, room] = busy.top(); busy.pop();
                busy.emplace(next_free_time + duration, room);
                ++count[room];
            }
        }

        int maxMeetings = 0, bestRoom = 0;
        for (int i = 0; i < n; ++i) {
            if (count[i] > maxMeetings) {
                maxMeetings = count[i];
                bestRoom = i;
            }
        }
        return bestRoom;
    }
};
